[svg] text transformation not starting on initial render

graouts opened the pull request:
#### 1fe8f8ec39961f73275ec569545709774886e2e5
<pre>
[svg] text transformation not starting on initial render
<a href="https://bugs.webkit.org/show_bug.cgi?id=253259">https://bugs.webkit.org/show_bug.cgi?id=253259</a>

Reviewed by Nikolas Zimmermann.

RenderSVGText has a custom implementation for localTransform() and localToParentTransform() which
require setNeedsTransformUpdate() to be called when the transform needs an update. This would not
be called correctly when the transform was changed via script.

We now have an override for styleDidChange() which looks over transform-related RenderStyle methods
to determine whether style has changed.

* LayoutTests/svg/text/text-transform-style-changed-by-script-expected.html: Added.
* LayoutTests/svg/text/text-transform-style-changed-by-script.html: Added.
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGText.h:

Canonical link: <a href="https://commits.webkit.org/261408@main">https://commits.webkit.org/261408@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76cbbba3af4b1a34dde40be8a858a2474c5e2b93

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111473 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20609 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/76 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3267 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120257 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115532 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21979 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11714 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2686 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117236 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16318 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99473 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104205 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98274 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/56 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45018 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13107 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/54 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86818 "run-api-tests-without-change (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13611 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9498 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19056 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52037 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15578 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4329 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -628,4 +628,30 @@ void RenderSVGText::updatePositionAndOverflow(const FloatRect& boundaries)
 
+void RenderSVGText::styleDidChange(StyleDifference diff, const RenderStyle* oldStyle)
+{
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
```
Correct! Not needed for LBSE.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -628,4 +628,30 @@ void RenderSVGText::updatePositionAndOverflow(const FloatRect& boundaries)
+        auto needsTransformUpdate = [oldStyle](const RenderStyle& newStyle) {
+            if (!oldStyle)
+                return newStyle.affectsTransform();
+            return (oldStyle->affectsTransform() != newStyle.affectsTransform()
```
I'd omit a) omit the outer paraenthesis and b) make this individual early returns instead of one long condition.

graouts replied:
I've wrapped the LBSE and `diff == StyleDifference::Layout` checks in the lambda so now it's just return statements and no long if clause.

nikolaszimmermann reviewed (approved):
LGTM, r=me with one nitpick.

graouts committed:
[svg] text transformation not starting on initial render
https://bugs.webkit.org/show_bug.cgi?id=253259

Reviewed by Nikolas Zimmermann.

RenderSVGText has a custom implementation for localTransform() and localToParentTransform() which
require setNeedsTransformUpdate() to be called when the transform needs an update. This would not
be called correctly when the transform was changed via script.

We now have an override for styleDidChange() which looks over transform-related RenderStyle methods
to determine whether style has changed.

* LayoutTests/svg/text/text-transform-style-changed-by-script-expected.html: Added.
* LayoutTests/svg/text/text-transform-style-changed-by-script.html: Added.
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGText.h:

Canonical link: https://commits.webkit.org/261408@main
